downloader/src/main/kotlin/VcsHost.kt

```diff
@@ -44,6 +46,57 @@ enum class VcsHost(
      */
     vararg supportedTypes: VcsType
 ) {
+    AZURE_DEVOPS("dev.azure.com", VcsType.GIT) {
+        private val gitCommitPrefix = "GC"
+
+        override fun getUserOrOrgInternal(projectUrl: URI) = projectUrlToUserOrOrgAndProject(projectUrl)?.first
+
+        override fun getProjectInternal(projectUrl: URI) = projectUrl.path.substringAfterLast("/")
+
+        override fun toVcsInfoInternal(projectUrl: URI): VcsInfo {
+            val uri = with(projectUrl) { URI(scheme, authority, path, null, fragment) }
+            val revision = projectUrl.getQueryParameters()["version"]?.firstOrNull()
+                .withoutPrefix(gitCommitPrefix).orEmpty()
+            val path = projectUrl.getQueryParameters()["path"]?.firstOrNull().withoutPrefix("/").orEmpty()
+
+            return VcsInfo(VcsType.GIT, uri.toString(), revision, path)
+        }
+
+        override fun toPermalinkInternal(vcsInfo: VcsInfo, startLine: Int, endLine: Int): String {
+            val actualEndLine = if (endLine != -1) endLine else startLine + 1
+
+            val lineQueryParam = "line=$startLine&lineEnd=$actualEndLine&lineColumn=1&lineEndColumn=1"
+            val pathQueryParam = if (vcsInfo.path.isNotEmpty()) "&path=/${vcsInfo.path}" else ""
+            val revisionQueryParam = "&version=$gitCommitPrefix${vcsInfo.revision}".takeUnless {
+                vcsInfo.revision.isEmpty()
+            }.orEmpty()
+            return "${vcsInfo.url}?$lineQueryParam$pathQueryParam$revisionQueryParam"
+        }
+
+        override fun toArchiveDownloadUrlInternal(userOrOrg: String, project: String, vcsInfo: VcsInfo): String {
+            val pathIterator = Paths.get(URI(vcsInfo.url).path).iterator()
+            val team = if (pathIterator.hasNext() && pathIterator.next().toString() == userOrOrg) {
+                if (pathIterator.hasNext()) pathIterator.next() else ""
+            } else ""
```

Comment:
Going a bit wild alternatively 😉 

            val team = pathIterator.nextOrNull().takeIf { it.toString() == userOrOrg }?.let {
                pathIterator.nextOrNull()?.toString()
            }.orEmpty()

Similar below.